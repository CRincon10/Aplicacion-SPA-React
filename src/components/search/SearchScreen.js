import {useNavigate, useLocation} from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from "../../hooks/useForm"
import {getHeroesByName} from '../../selectors/getHeroesByName'
import {HeroCard} from '../hero/HeroCards'
import { useMemo } from 'react'

export const SearchScreen = () => {

    const navigate = useNavigate()
    const location = useLocation()

    const query = queryString.parse(location.search)
    const {q = '' } = query       //solo estoy desestructurando la prop q de query y le di valor por default en caso de que no tenga ningun valor inicial


    const [ {searchText},handleInputChange] = useForm({
        searchText:q                     //initialState asi es como se va a ver el input de forma inicial en este caso va a mantener el valor de la busqueda ya que ese es el valor que tiene q
    })
    
    const heroesfilter = useMemo(()=> getHeroesByName(q), [q])

    const handleSearchText = (e)=>{
        e.preventDefault()

        if(searchText.trim().length <=2 ){
            return;
        }

        navigate(`?q=${searchText}`)  //metodo de busqueda que va a dirigirme a la ruta que se le indique dentro del ${}, en este punto ya navegue hacia esa ruta lo puedo comprobar en el url del navegador
    
    }

    return (
        <>
            <h1>Your Hero</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr/>

                    <form onSubmit={handleSearchText}>
                        <input 
                            type='text'
                            placeholder="Find your hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={handleInputChange}

                        />
                        <button
                            type="submit"
                            className="btn btn-outline-secondary mt-3"
                        >
                            search...
                        </button>

                    </form>

                </div>

                <div className='col-7'>
                        <h4>Results</h4>
                        <hr/>

                        {
                            (q === '')
                                ? <div className='alert alert-info'>look for a hero</div>
                                : (heroesfilter.length === 0 ) && <div className='alert alert-danger'> No results found for: {q} </div>
                        }

                        {
                            heroesfilter.map( hero =>( 
                                <HeroCard 
                                    key={hero.id}
                                    {...hero}

                                />
                            ))
                        }

                </div>

            </div>

        </>
    )
}





/*
En este componente use el hook useForm que me retorna el values que es el initialState el handleInputChange que retorna 
los values que tenga pero al target.name le computa el valor del target.value y el reset para reestablecer el valor del input

query parameters desde la busqueda
El handleSearchText retorna el searchText y ese va a ser el valor de la busqueda lo que voy a hacer es navegar a la ruta donde el queryParameter va a estar establecido
navigate(`?q=${searchText}`)  ==> metodo de busqueda que va a dirigirme a la ruta que se le indique dentro del ${}
useLocation  == > hook que indica la localizacion donde nos encontramos dentro de la app tiene varias propiedades la que uso en este componente es search
es necesario usar query-string instalacion de npm install query-string e importarlo 
const query = queryString.parse(location.search)    ==> a queryString le parseo el valor de location con su propiedad serch


*/