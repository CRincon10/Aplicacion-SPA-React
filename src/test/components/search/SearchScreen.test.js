import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom"
import { SearchScreen } from "../../../components/search/SearchScreen"



describe('Test en <SearchScreen/>', () => { 

    

    test('Debe mostrarse correctamente y con valores por defecto', () => {  

        const wrapper = mount(
                            <MemoryRouter initialEntries={['/search']}>
                                <SearchScreen/>
                            </MemoryRouter>
                            )

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').text().trim()).toBe('look for a hero')
    })

    test('Debe mostrar a batman y el input con el valor del queryString', () => { 

        const wrapper = mount(
                            <MemoryRouter initialEntries={['/search?q=batman']}>
                                <SearchScreen/>
                            </MemoryRouter>
                            )
        
        expect(wrapper.find('.alert-info').length).toBe(0)
        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe mostrar el error si no se encuentra el Hero', () => {  

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1234']}>
                <SearchScreen/>
            </MemoryRouter>
            )

        
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No results found for: batman1234`)
        


    })

})







/*
Para hacer pruebas con componentes que tengan el router y tambien url necesito proveer un contexto en el cual hayan rutas es decir al <SearchScreen/> es necesario
proveerle el useNavigate() y un contexto que tenga rutas por eso es necesario importar el MemoryRouter que es el que me permite hacer evaluaciones y pruebas como si 
estuviera en el navegador ese <MemoryRouter/> debe estar dentro del context y debe contener el componente que estamos evaluando par simular las rutas, tambiÃ©n es 
necesario que le indique el path de la ruta que vamos a evaluar ya que depende de esto me va a renderizar una ruta o componente, se llama el initialEntries y dentro 
de este el segmento de url en el cual definimos el path initialEntries={['/search']}

Los queryParemeters eventualmente son leidos por 
const query = queryString.parse(location.search)
basados en esos queryparameters se extrae el q   const {q = '' } = query  
basado en el q se construye el valor que tiene el texto
 const [ {searchText},handleInputChange] = useForm({
    searchText:q                     
})
basado en esto se muestran los heroes que coincidan con ese searchText
const heroesfilter = useMemo(()=> getHeroesByName(q), [q])
Entonces si yo tengo un valor el componente reacciona o se renderiza de alguna manera de forma condicional



Test:
1.Debe mostrarse correctamente y con valores por defecto
2.Debe mostrar a batman y el input con el valor del queryString
  initialEntries={['/search?q=batman']}  lo defino de esta forma porque esta es la ruta o path para que muestre a batman
*/
