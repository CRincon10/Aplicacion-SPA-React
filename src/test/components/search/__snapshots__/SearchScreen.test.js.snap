// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en <SearchScreen/> Debe mostrar a batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Your Hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-secondary mt-3"
          type="submit"
        >
          search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-tittle"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Test en <SearchScreen/> Debe mostrarse correctamente y con valores por defecto 1`] = `
Array [
  <h1>
    Your Hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-secondary mt-3"
          type="submit"
        >
          search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        look for a hero
      </div>
    </div>
  </div>,
]
`;
